src/shared/lib/PdbUtils.ts

```diff
@@ -153,69 +131,80 @@ function groupAlignmentsByPdbIdAndChain(alignments: PdbUniprotAlignment[])
-        end = alignments[0].uniprotFrom - 1;
+        start = alignments[0].uniprotFrom;
     }
     else {
         return undefined;
     }
 
     alignments.forEach((alignment: PdbUniprotAlignment) => {
-        const distance = alignment.uniprotFrom - end - 1;
         const alignmentStr = generateAlignmentString(alignment) || "";
+        let diff = Math.abs(alignment.uniprotFrom - start);
 
-        // check for overlapping uniprot positions...
+        if (alignment.uniprotFrom < start) {
+            let gapOrOverlap = Math.abs(alignmentStr.length - diff);
 
-        // no overlap, and the next alignment starts exactly after the current merge
-        if (distance === 0) {
-            // just concatenate two strings
-            mergedAlignment += alignmentStr;
+            // overlap
+            if (alignmentStr.length >= diff) {
+                mergedAlignment = alignmentStr + mergedAlignment.substr(gapOrOverlap);
+            }
+            // gap
+            else {
+                mergedAlignment = alignmentStr + alignmentGap(gapOrOverlap) + mergedAlignment;
+            }
         }
-        // no overlap, but there is a gap
-        else if (distance > 0)
-        {
-            let gap: string[] = [];
+        else if (alignment.uniprotFrom >= start) {
+            let gapOrOverlap = Math.abs(mergedAlignment.length - diff);
 
-            // add gap characters (character count = distance)
-            for (let i = 0; i < distance; i++) {
-                gap.push(ALIGNMENT_GAP);
+            // overlap
+            if (mergedAlignment.length >= diff) {
+                mergedAlignment = mergedAlignment.substr(0, mergedAlignment.length - gapOrOverlap) + alignmentStr +
+                    mergedAlignment.substr(mergedAlignment.length - gapOrOverlap + alignmentStr.length);
+            }
+            // gap
+            else {
+                mergedAlignment = mergedAlignment + alignmentGap(gapOrOverlap) + alignmentStr;
             }
-
-            // also add the actual string
-            mergedAlignment += gap.join("") + alignmentStr;
-        }
-        // overlapping
-        else {
-            // TODO: here we are assuming that overlapping sections are the same for both alignment strings,
-            // but that's not always the case...
-
-            // merge two strings
-            // (just append the non-overlapping section of the current alignment)
-            mergedAlignment += alignmentStr.substr(-1 * distance);
         }
 
-        // update the end position
-        end = Math.max(end, alignment.uniprotTo);
+        start = Math.min(start, alignment.uniprotFrom);
     });
 
     return {
         pdbId: alignments[0].pdbId,
         chain: alignments[0].chain,
-        uniprotStart: alignments[0].uniprotFrom,
-        uniprotEnd: alignments[0].uniprotFrom + mergedAlignment.length,
+        uniprotStart: start,
+        uniprotEnd: start + mergedAlignment.length - 1,
```

Comment:
maybe it would be good to document this, that `uniprotEnd` is inclusive.  Avoid some off-by-one errors :)